### PR TITLE
Add config to check BenevolentUnion

### PR DESCRIPTION
**Pull request by VincentLanglet**
Closes https://github.com/phpstan/phpstan/issues/8223

### CODE REVIEW

File: conf/config.neon [hunk 1]
```diff
@@ -76,6 +76,7 @@ parameters:
 	checkNullables: false
 	checkThisOnly: true
 	checkUnionTypes: false
+	checkBenevolentUnionTypes: false
```
Comment (staabm) [inline conf/config.neon:79]: should it be true on bleedingEdge?

Comment (VincentLanglet) [reply conf/config.neon:79]: CheckUnionTypes is not true on bleedingEdge, so checkBenevolentUnionTypes which is stricter shouldn't be true too.

Comment (ondrejmirtes) [reply conf/config.neon:79]: `checkBenevolentUnionTypes` will never be true by default, that's why it's "benevolent".

File: src/Rules/RuleLevelHelper.php [hunk 2]
```diff
@@ -48,6 +49,16 @@ public function isThis(Expr $expression): bool
 	/** @api */
 	public function accepts(Type $acceptingType, Type $acceptedType, bool $strictTypes): bool
 	{
+		$checkForUnion = $this->checkUnionTypes;
+
+		if (
+			$this->checkBenevolentUnionTypes
+			&& $acceptedType instanceof BenevolentUnionType
```
Comment (ondrejmirtes) [inline src/Rules/RuleLevelHelper.php]: This is too naive - will not see nested BenevolentUnionType. Needs similar logic as `checkExplicitMixed` and use `TypeTraverser`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Rules/RuleLevelHelper.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Rules/RuleLevelHelper.php]: I have a test for (\DateTimeImmutable|false) and for array<(\DateTimeImmutable|false)>.

What do you have in mind with "nested BenevolentUnionType", I'll add new tests.

Comment (staabm) [reply src/Rules/RuleLevelHelper.php]: I guess things like `(float|int)|string`

File: conf/config.neon (hunk 1 above)
Comment (VincentLanglet) [reply conf/config.neon:79]: But I'm kinda worried about someone using the config
```
checkUnionTypes: false
checkBenevolentUnionTypes: true
```
what should be the behavior ?
- checking both ?
- checking none ?
- only checking the benevolent ?
- throwing an error ?

**Comment by ondrejmirtes** [comment]
I think it's fine as it is. Please also send a documentation update to https://phpstan.org/config-reference (https://github.com/phpstan/phpstan/edit/1.9.x/website/src/config-reference.md).

Thank you.

**Comment by VincentLanglet** [comment]
> I think it's fine as it is. Please also send a documentation update to https://phpstan.org/config-reference (https://github.com/phpstan/phpstan/edit/1.9.x/website/src/config-reference.md).
> 
> Thank you.

Sure, I started https://github.com/phpstan/phpstan/pull/8234

**Pull request merged**